chainindex/events.go

```diff
@@ -0,0 +1,501 @@
+package chainindex
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	amt4 "github.com/filecoin-project/go-amt-ipld/v4"
+	"github.com/filecoin-project/go-state-types/abi"
+	blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type executedMessage struct {
+	msg types.ChainMsg
+	rct types.MessageReceipt
+	// events extracted from receipt
+	evs []types.Event
+}
+
+// events are indexed against their inclusion/message tipset when we get the corresponding execution tipset
+func (si *SqliteIndexer) indexEvents(ctx context.Context, tx *sql.Tx, msgTs *types.TipSet, executionTs *types.TipSet) error {
+	// check if we have an event indexed for any message in the `msgTs` tipset -> if so, there's nothig to do here
+	// this makes event inserts idempotent
+	msgTsKeyCidBytes, err := toTipsetKeyCidBytes(msgTs)
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// if we've already indexed events for this tipset, mark them as unreverted and return
+	res, err := tx.Stmt(si.updateEventsToNonRevertedStmt).ExecContext(ctx, msgTsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("failed to unrevert events for tipset: %w", err)
+	}
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("failed to get rows affected by unreverting events for tipset: %w", err)
+	}
+	if rows > 0 {
+		log.Infof("unreverted %d events for tipset: %s", rows, msgTs.Key())
+		return nil
+	}
+
+	if !si.cs.IsStoringEvents() {
```

Comment:
The placement of this so far down in the function is rational, but I wonder if it's worth the bother. It means that `updateEventsToNonReverted` is execute for every tipset regardless of whether we have new events or not. In the common case it'd be on an empty table, so nice and quick, but still entirely unnecessary. All to handle the case of "I once had this turned on but then turned it off". But is that a case we care about? You're going to have problems if you do that anyway.

Reply:
I am okay with going the extra mile here to not compromise on data integrity. If we move this to the top of this function, all we really save is one DB read query in the very uncommon case of "I want to Index stuff, just not the events".

In the case of people who are indexing everything, there is no additional cost here.